examples/infer/bfs.desc

```diff
@@ -1,125 +1,114 @@
+    mask: bool,
+    updatingMask: bool,
+    visited: bool,
+    cost: T
+}
+
+struct Node {
+    start: i32,
+    number_ofEdges: i32
+}
+impl Copy for Node {}
+
+trait Incrementable: Copy {
+    fn increment(Self) -[gpu.thread]-> Self;
+}
+
+impl Incrementable for i32 {
+    fn increment(self) -[gpu.thread]-> Self {
+        self + 1
     }
 }
 
-fn iteration<r1: prv, j: prv , k: prv>(
-    nod2: &r1 shrd gpu.global (i32, i32),
-    chgn2: &j uniq gpu.global (bool, bool, bool, i32), // mask, updating mask, visited, cost
-    ed2: &k shrd gpu.global [i32; m]
-) -[gpu.thread]-> () <'c1, 'c2, 'c3>{
-    // let mut change = *chgn2;
+fn iteration<m: nat, T>(
+    nod2: &shrd gpu.global Node,
+    chgn2: &uniq gpu.global HChange<T>,
+    ed2: &shrd gpu.global [i32; m]
+) -[gpu.thread]-> () where T: Incrementable {
     let nodes2 = *nod2;
-    if (*chgn2).0 {
-
-        (*chgn2).0 = false;
-        let node_start = nodes2.0;
-        let no_of_edges = nodes2.1;
-        let upper_range = node_start+no_of_edges;
-        for i in node_start..upper_range <'c3>{
-            let index = ed2[i];
-
-            let change_ptr = to_raw_ptr::<'c3, gpu.global, (bool, bool, bool, i32)>(&'c3 uniq *chgn2);
-            let ch_ptr_i = offset_raw_ptr::<(bool, bool, bool, i32)>(change_ptr, index);
-            // let mut child = *ch_ptr_i;
-
-            let was_visited = (*ch_ptr_i).2;
-            if (was_visited == false) {
-                // atomic_set::<'c3 , gpu.global, i32>( &'c3 uniq (*ch_ptr_i).3 ,(*chgn2).3 + 1);
-                (*ch_ptr_i).3 = (*chgn2).3 + 1; // cost
-                (*ch_ptr_i).1 = true // updating mask
-            } else {
-                ()
+    if (*chgn2).mask {
+        (*chgn2).mask = false;
+
+        let upper_range = nodes2.start + nodes2.number_ofEdges;
+        for i in nodes2.start..upper_range {
+            let change_ptr = to_raw_ptr(&uniq *chgn2);
+            let ch_ptr_i = offset_raw_ptr(change_ptr, ed2[i]);
+
+            if !(*ch_ptr_i).visited {
+                (*ch_ptr_i).cost = (*chgn2).cost.increment();
+                (*ch_ptr_i).updatingMask = true
             }
         }
-    } else {
-        ()
     }
 }
 
-fn bfs<n:nat, m:nat, a: prv, b: prv, k:prv, l:prv, a1:prv>(
-    h_nodes: &a shrd cpu.heap [(i32, i32); n],
-    h_edges: &b shrd cpu.heap [i32; m],
-    h_changes: &k uniq cpu.heap [(bool, bool, bool, i32); n],
-    h_stop: &a1 uniq cpu.heap bool
-) -[cpu.thread]-> () <'c, 'd, 'e, 'f, 'g, 'h, 'a2, 'a3, 'a4>{
+fn bfs<n:nat, m:nat>(
+    h_nodes: &shrd cpu.mem [Node; n],
+    h_edges: &shrd cpu.mem [i32; m],
+    h_changes: &uniq cpu.mem [HChange<i32>; n],
+    h_stop: &uniq cpu.mem bool
+) -[cpu.thread]-> () {
+    let prepare_iteration = |change3, stop3| -[gpu.thread]-> () {
+        if (*change3).updatingMask {
+            (*change3).mask = true;
+            (*change3).visited = true;
+            atomic_set(stop3, true);
+            (*change3).updatingMask = false
+        }
+    };
```

Comment:
This can now be written as lambda-function without type annotations.
The types are:
```Rust
    change3: &uniq gpu.global HChange<i32>,
    stop3: &uniq gpu.global bool
```
They can now be inferred automatically.
There was the following problem: if we saw a deref of an implicit type-identifier $T we replaced $T by a reference. At this point we had to make the decision if this is an shrd-reference or an uniq-reference. (RawPtr were ignored and could not be inferred).
Now we simply assume this is a type which implements the deref-Trait and ignore borrowing-checks. In a second pass, when all type are inferred, we execute normal typechecking with borrowing-checks.